src/nanoemoji/parts.py

```diff
@@ -89,12 +123,82 @@ def add(self, source: PathSource):
                 svg_shape: SVGShape = a_source  # pytype: disable=attribute-error
                 self._add(Shape(svg_shape.as_path().d))
 
+    def _compute_donor(self, norm: NormalizedShape):
+        self._donor_cache[norm] = None  # no solution
+
+        # try to select a donor that can fulfil every member of the set
+        # the input shape is in the set so if found we can get from donor => input
+        # shrinking a big thing is more likely to result in small #s that fit into
+        # more compact PaintTransform variants so try biggest first
+
+        # TODO there are cases where this picks a suboptimal transform, e.g. a 2x3
+        # downscale be used when a scale uniform around center upscale might work
+        # Ex SVGPath(d="M8,13 A3 3 0 1 1 2,13 A3 3 0 1 1 8,13 Z")
+        #    SVGPath(d="M11,5 A2 2 0 1 1 7,5 A2 2 0 1 1 11,5 Z")
+        # A fancier implementation would factor in the # of occurences and the cost
+        # based on which shape is selected as donor if there are many possibilities.
+
+        svg_paths = sorted(
+            self.shape_sets[norm], key=lambda s: (_bbox_area(s), s), reverse=True
+        )
+        svg_paths = [SVGPath(d=s) for s in svg_paths]
+
+        for svg_path in svg_paths:
+            if all(
+                affine_between(svg_path, svg_path2, self.reuse_tolerance) is not None
+                for svg_path2 in svg_paths
```

Comment:
in this nested for loops, should you not skip the case when `svg_path is svg_path2`, which is guaranteed to be identity? In itertools term I think you want `permutations` instead of `product`

Reply:
I could but it adds a condition that shouldn't really matter, affine_between should handle it.